pkg/gateway/gateway.go

```diff
@@ -0,0 +1,165 @@
+package gateway
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/pkg"
+	"github.com/threefoldtech/zos/pkg/network/namespace"
+	"github.com/threefoldtech/zos/pkg/network/types"
+	"github.com/threefoldtech/zos/pkg/zinit"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	traefikService = "traefik"
+)
+
+type gatewayModule struct {
+	proxyConfigPath string
+	traefikStarted  bool
+}
+
+type ProxyConfig struct {
+	Http HTTPConfig
+}
+
+type HTTPConfig struct {
+	Routers  map[string]Router
+	Services map[string]Service
+}
+
+type Router struct {
+	Rule    string
+	Service string
+}
+
+type Service struct {
+	LoadBalancer LoadBalancer
+}
+
+type LoadBalancer struct {
+	Servers []Server
+}
+
+type Server struct {
+	Url string
+}
+
+func New(root string) pkg.Gateway {
+	configPath := filepath.Join(root, "proxy")
+	// where should service-restart/node-reboot recovery be handled?
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
+		err := os.MkdirAll(configPath, 0644)
```

Comment:
MkdirAll is safe to call even if directory exists. the checks if it exists is a nice touch, but not needed.